File: web-ui/tests/unit/wakeRouting.test.js
import { describe, expect, it } from "vitest";
import { createSign, generateKeyPairSync } from "node:crypto";

import {
  bridgeCheckinEventId,
  describeWakeRouting,
} from "../../src/lib/wakeRouting.js";

const bridgeProofKey = (() => {
  const { publicKey, privateKey } = generateKeyPairSync("ec", {
    namedCurve: "prime256v1",
    publicKeyEncoding: { format: "der", type: "spki" },
    privateKeyEncoding: { format: "der", type: "pkcs8" },
  });
  return {
    privateKey,
    publicKeyB64: Buffer.from(publicKey).toString("base64"),
  };
})();

function stableJsonValue(value) {
  if (Array.isArray(value)) {
    return value.map((item) => stableJsonValue(item));
  }
  if (value && typeof value === "object") {
    return Object.keys(value)
      .sort()
      .reduce((normalized, key) => {
        normalized[key] = stableJsonValue(value[key]);
        return normalized;
      }, {});
  }
  return value;
}

function signCheckinPayload(content) {
  const signer = createSign("sha256");
  signer.update(
    Buffer.from(
      JSON.stringify(
        stableJsonValue({
          v: "agent-bridge-checkin-proof/v1",
          handle: String(content.handle ?? "").trim(),
          actor_id: String(content.actor_id ?? "").trim(),
          workspace_id: String(content.workspace_id ?? "").trim(),
          bridge_instance_id: String(content.bridge_instance_id ?? "").trim(),
          checked_in_at: String(content.checked_in_at ?? "").trim(),
          expires_at: String(content.expires_at ?? "").trim(),
        }),
      ),
      "utf8",
    ),
  );
  signer.end();
  return signer
    .sign({ key: bridgeProofKey.privateKey, format: "der", type: "pkcs8" })
    .toString("base64");
}

async function checkinEvent(content) {
  return {
    state: "ok",
    document: {
      event: {
        id: "event-bridge-checkin-1",
        type: "agent_bridge_checked_in",
        payload: {
          ...content,
          proof_signature_b64: await signCheckinPayload(content),
        },
      },
    },
  };
}

describe("wakeRouting", () => {
  const principal = {
    principal_kind: "agent",
    actor_id: "actor-ops-ai",
    username: "m4-hermes",
    revoked: false,
  };

  it("marks an agent online when the durable workspace id is bound", async () => {
    const { publicKeyB64 } = bridgeProofKey;
    const result = await describeWakeRouting(
      {
        ...principal,
        registration: {
          handle: "m4-hermes",
          actor_id: "actor-ops-ai",
          status: "active",
          bridge_signing_public_key_spki_b64: publicKeyB64,
          bridge_checkin_event_id: "event-bridge-checkin-1",
          workspace_bindings: [{ workspace_id: "ws-123", enabled: true }],
        },
      },
      null,
      "ws-123",
      await checkinEvent({
        handle: "m4-hermes",
        actor_id: "actor-ops-ai",
        workspace_id: "ws-123",
        bridge_instance_id: "bridge-hermes-1",
        checked_in_at: "2099-03-01T10:00:00Z",
        expires_at: "2099-03-01T10:05:00Z",
      }),
    );

    expect(
      bridgeCheckinEventId({
        ...principal,
        registration: {
          handle: "m4-hermes",
          actor_id: "actor-ops-ai",
          bridge_checkin_event_id: "event-bridge-checkin-1",
        },
      }),
    ).toBe("event-bridge-checkin-1");
    expect(result).toMatchObject({
      applicable: true,
      taggable: true,
      online: true,
      offline: false,
      state: "online",
      badgeLabel: "Online",
      summary: "Online as @m4-hermes.",
    });
  });

  it("distinguishes transient registration lookup failures from missing docs", async () => {
    await expect(
      describeWakeRouting(principal, { state: "missing" }, "ws-123", null),
    ).resolves.toMatchObject({
      applicable: true,
      taggable: false,
      online: false,
      offline: false,
      state: "unregistered",
      badgeLabel: "Unregistered",
      summary: "Missing wake registration for @m4-hermes.",
    });

    await expect(
      describeWakeRouting(principal, { state: "error" }, "ws-123", null),
    ).resolves.toMatchObject({
      applicable: true,
      taggable: false,
      online: false,
      offline: false,
      state: "unknown",
      badgeLabel: "Unknown",
      summary: "Registration status is unavailable right now.",
    });

    const { publicKeyB64 } = bridgeProofKey;
    await expect(
      describeWakeRouting(
        {
          ...principal,
          registration: {
            handle: "m4-hermes",
            actor_id: "actor-ops-ai",
            status: "active",
            bridge_signing_public_key_spki_b64: publicKeyB64,
            bridge_checkin_event_id: "event-bridge-checkin-1",
            workspace_bindings: [{ workspace_id: "ws-123", enabled: true }],
          },
        },
        null,
        "ws-123",
        { state: "error" },
      ),
    ).resolves.toMatchObject({
      applicable: true,
      taggable: true,
      online: false,
      offline: true,
      state: "unknown",
      badgeLabel: "Unknown",
      summary: "Bridge check-in status is unavailable right now.",
    });
  });

  it("keeps healthy bridge registrations online when page workspace context is missing", async () => {
    const { publicKeyB64 } = bridgeProofKey;
    const result = await describeWakeRouting(
      {
        ...principal,
        registration: {
          handle: "m4-hermes",
          actor_id: "actor-ops-ai",
          status: "active",
          bridge_signing_public_key_spki_b64: publicKeyB64,
          bridge_checkin_event_id: "event-bridge-checkin-1",
          workspace_bindings: [{ workspace_id: "ws-123", enabled: true }],
        },
      },
      null,
      "",
      await checkinEvent({
        handle: "m4-hermes",
        actor_id: "actor-ops-ai",
        workspace_id: "ws-123",
        bridge_instance_id: "bridge-hermes-1",
        checked_in_at: "2099-03-01T10:00:00Z",
        expires_at: "2099-03-01T10:05:00Z",
      }),
    );

    expect(result).toMatchObject({
      applicable: true,
      taggable: true,
      online: true,
      offline: false,
      state: "online",
      badgeLabel: "Online",
      summary:
        "Online for bound workspace ws-123, but this page has no durable workspace ID to confirm the current workspace match.",
    });
  });

  it("keeps registered agents taggable but offline until the bridge checks in", async () => {
    const { publicKeyB64 } = bridgeProofKey;
    const result = await describeWakeRouting(
      {
        ...principal,
        registration: {
          handle: "m4-hermes",
          actor_id: "actor-ops-ai",
          status: "pending",
          bridge_signing_public_key_spki_b64: publicKeyB64,
          workspace_bindings: [{ workspace_id: "ws-123", enabled: true }],
        },
      },
      null,
      "ws-123",
      null,
    );

    expect(result).toMatchObject({
      applicable: true,
      taggable: true,
      online: false,
      offline: true,
      state: "offline",
      badgeLabel: "Offline",
      summary:
        "Offline. The agent is registered for this workspace, but no fresh bridge check-in is available yet.",
    });
  });

  it("treats stale bridge check-ins as offline", async () => {
    const { publicKeyB64 } = bridgeProofKey;
    const result = await describeWakeRouting(
      {
        ...principal,
        registration: {
          handle: "m4-hermes",
          actor_id: "actor-ops-ai",
          status: "active",
          bridge_signing_public_key_spki_b64: publicKeyB64,
          bridge_checkin_event_id: "event-bridge-checkin-1",
          workspace_bindings: [{ workspace_id: "ws-123", enabled: true }],
        },
      },
      null,
      "ws-123",
      await checkinEvent({
        handle: "m4-hermes",
        actor_id: "actor-ops-ai",
        workspace_id: "ws-123",
        bridge_instance_id: "bridge-hermes-1",
        checked_in_at: "2026-03-01T10:00:00Z",
        expires_at: "2026-03-01T10:05:00Z",
      }),
    );

    expect(result).toMatchObject({
      applicable: true,
      taggable: true,
      online: false,
      offline: true,
      state: "offline",
      badgeLabel: "Offline",
      summary:
        "Offline. The agent is registered for this workspace, but its last bridge check-in is stale.",
    });
  });
});

File: core/internal/auth/wake_routing_test.go
package auth

import (
	"testing"
	"time"
)

func TestDescribeWakeRoutingMarksFreshHeartbeatOnline(t *testing.T) {
	t.Parallel()

	status := DescribeWakeRouting(
		AuthPrincipalSummary{
			ActorID:       "actor-m4-hermes",
			Username:      "m4-hermes",
			PrincipalKind: string(PrincipalKindAgent),
			Registration: &AgentRegistration{
				Handle:            "m4-hermes",
				ActorID:           "actor-m4-hermes",
				Status:            "active",
				BridgeInstanceID:  "bridge-hermes-1",
				BridgeCheckedInAt: "2099-03-20T12:00:00Z",
				BridgeExpiresAt:   "2099-03-20T12:05:00Z",
				WorkspaceBindings: []AgentRegistrationWorkspaceBinding{
					{WorkspaceID: "ws_main", Enabled: true},
				},
			},
		},
		"ws_main",
		time.Date(2099, 3, 20, 12, 1, 0, 0, time.UTC),
	)

	if !status.Applicable || !status.Taggable || !status.Online {
		t.Fatalf("expected wake routing to be online and taggable, got %#v", status)
	}
	if status.State != WakeRoutingStateOnline {
		t.Fatalf("expected online state, got %#v", status)
	}
	if status.Summary != "Online as @m4-hermes." {
		t.Fatalf("unexpected online summary: %#v", status)
	}
}

func TestDescribeWakeRoutingTreatsStaleHeartbeatAsOffline(t *testing.T) {
	t.Parallel()

	status := DescribeWakeRouting(
		AuthPrincipalSummary{
			ActorID:       "actor-m4-hermes",
			Username:      "m4-hermes",
			PrincipalKind: string(PrincipalKindAgent),
			Registration: &AgentRegistration{
				Handle:            "m4-hermes",
				ActorID:           "actor-m4-hermes",
				Status:            "active",
				BridgeInstanceID:  "bridge-hermes-1",
				BridgeCheckedInAt: "2026-03-20T12:00:00Z",
				BridgeExpiresAt:   "2026-03-20T12:05:00Z",
				WorkspaceBindings: []AgentRegistrationWorkspaceBinding{
					{WorkspaceID: "ws_main", Enabled: true},
				},
			},
		},
		"ws_main",
		time.Date(2026, 3, 20, 12, 6, 0, 0, time.UTC),
	)

	if !status.Applicable || !status.Taggable || status.Online {
		t.Fatalf("expected wake routing to stay taggable but offline, got %#v", status)
	}
	if status.State != WakeRoutingStateOffline {
		t.Fatalf("expected offline state, got %#v", status)
	}
	if status.Summary != "Offline. The agent is registered for this workspace, but its last bridge heartbeat is stale." {
		t.Fatalf("unexpected offline summary: %#v", status)
	}
}
